### ISSUE TITLE
The need for `Set`

### ISSUE THREAD

**Issue by anshumanmohan** [cucapra/pollen#64]
I think https://github.com/cucapra/pollen/issues/57#issuecomment-1516698587 raises an interesting question about how to generate steps for an output graph. We can force users to simply add steps to an ordinary `steps': Set<Step>`, and then just emit `{ node with steps = steps'}` for each node in the graph. I don't know of any odgi functions which compute each element of `steps'` in parallel, so perhaps this solution is sufficient for most graph analyses. I think the alternative would be to allow users emit steps to `step'` and then also use `step'` as a first-class value, but as discussed today this is more of a stretch goal and I think we should stick with forcing `steps'` to be an ordinary `Set` of steps.

Of course, to fully support this, we probably want to extend pollen with support for creating and building `Set`s.

_Originally posted by @susan-garry in https://github.com/cucapra/pollen/issues/57#issuecomment-1517097131_

**Comment by anshumanmohan** [comment]
When I drew up the little Pollen example where I `emit`ted steps of a node, I was sorta just staying close the motivating example @susan-garry had going in the main body of #57.

In general, I thought that the steps of a node are precomputed for the user's benefit, and are available for iterating over but not for modification. Does the user ever really get to modify the steps of a node directly?

**Comment by sampsyo** [comment]
To summarize, I think the status quo is that we're designing Pollen (1.0) to allow the _write-only_ construction of output sets. That is, our `emit` statements let you put stuff _into_ output sets, but you can never _read from_ output sets (e.g., test for membership in them, or iterate over them). This means our programs are all "one-shot" set constructors. This issue, I think, asks whether we ever want to do something more complicated than that: e.g., construct a set (in parallel) and then use that set to construct another one (also in parallel).

To record my position on this that I aired synchronously: let's not worry about this for now. I think it is a vast and useful simplification to do the one-shot, write-only thing. We may want to relax this in the future, but it's going to be tricky, and we have plenty of work to do to get the one-shot stuff working how we want to first—and everything we do to get that right will help us do the multi-shot thing in the future well too (so it won't be wasted work).

**Comment by susan-garry** [comment]
Part of the issue is if, in the future, we might want to construct, say, `steps'` in parallel and then write `emit { node with steps = steps'`. I think the core issue, however, has to do with the semantics of generating a new graph.

Our type interface must maintain certain invariants, such that if there exists a path that crosses from node A to node B, then A and B must each contain steps that reflect this. If a user wants to generate a graph that contains a new path, do they need to emit `out_paths` and `out_nodes` with `out_node.steps` being modified where appropriate? Or if a user emits only `out_paths`, is this an error because `out_nodes` is empty or do we infer what `out_nodes` should be and essentially regenerate `graph.nodes` and `node.steps` to be consistent with `out_paths`? What if a user emits `out_paths` and `out_nodes` and they contain contradictory information about which paths cross which nodes? Should we allow users to emit `out_nodes` at all?

**Comment by sampsyo** [comment]
Sounds interesting! Correct me if I'm wrong here: it seems like the issue of consistency (i.e., maintaining invariants in the output) could be orthogonal from the issue of write-only sets (i.e., my suggestion that we stick to one-shot set construction for now). That is, even if we are in the write-only world of single-shot programs, it could be possible to construct inconsistent output graphs. Is that true?
